### DISCUSSION TITLE
Helm Chart

### DISCUSSION THREAD

**Discussion by recharte** [openeverest/openeverest#653] [General]
Hi everyone,

We're excited to announce that we're moving forward with adding an official Helm chart for Percona Everest! 🎉 This will make deploying and managing Everest in Kubernetes even easier. To give you an early look, we've created a Proof of Concept (PoC) for the Helm chart, which you can check out [here](https://github.com/mayankshah1607/everest-helm-chart/).

Please note that this is a real PoC—it's unpolished, lacks proper documentation, and contains some hacky scripts just to get things working. Full development hasn't started yet, and we want to involve you in shaping it from the ground up.

## How can you help?
We’d love for you to **try out the PoC** and share your thoughts to help guide the next steps:
* How do you plan to use the Helm chart for Percona Everest in your Kubernetes environment?
* What key features or configurations would you find most valuable?
* Are there any advanced use cases or edge cases we should consider?
* How should we handle upgrades, rollbacks, or versioning to best suit your workflows?

Your feedback will be instrumental in shaping the development of the Helm chart. We look forward to hearing your insights and experiences with the PoC!

Cheers

**Comment by icc-garciaju** [comment]
Having a helm chart would allow automated and reproducible deployments.

Also it can be included as dependency for other charts as a bundle to initialize a full-featured cluster.

It looks like percona-monitoring is already included, thats good.

Also, being able to define backup repositories from the chart would be nice.

**Comment by travisghansen** [comment]
Great! Been waiting for this for a while. Couple considerations:

- we manage *everything* using argocd..by extension the use of helm hooks can be problematic so avoiding those would be great if applicable
- I see some resources in the repo that seem to reference olm..is it required to be running olm to use the chart?
- Is it possible to configure a watch all namespace setup for all the different flavors?

Thanks!

**Reply by mayankshah1607** [reply]
Hi @travisghansen, thanks for your feedback!

To address your first two points: OLM is an integral part of Everest's architecture, and we primarily rely on it to streamline the installation and upgrade processes for Percona operators, making it a necessary dependency. Due to the way OLM functions, certain imperative actions are required during installation, which is why we currently use Helm chart hooks to handle these one-time tasks. I understand that this adds some complexity, but I want to reassure you that we are planning to re-architect our installation process in a future release. Our goal is to make it more compatible with Helm and GitOps, ultimately phasing out the OLM artifacts and most of the hooks from the Helm chart.

> Is it possible to configure a watch all namespace setup for all the different flavors?

In short, no, we currently don't support that setup. This was an intentional architectural decision based on stability and security considerations. By isolating the operator(s) to a specific namespace, we limit the impact of any issues to the databases within that namespace, rather than affecting all databases across the cluster. This also reduces the blast radius in case of any problems. It also enables a phased rollout of new operator versions—you can test a new version in one namespace first, and only proceed with upgrading other instances after successful testing.

With that said, there's also no technical restriction that stops us from implementing such a setup, and we can certainly look into this. :)

**Reply by travisghansen** [reply]
Thanks for the consideration! That all sounds great and I will continue to monitor the progress!

**Reply by travisghansen** [reply]
For what it’s worth on the blast radius front, we do this by rolling out updates through entire environments before ever hitting production clusters, but the intent is not lost on me.

Our approach to self-service for individual teams (which operate generally autonomously in namespaces) really fits the cluster-wide model much better…as cluster admins we offer the menu of items and they chose what they need, requiring them (or me) to also install the operator on a per-ns basis is a pretty cumbersome ask in our workflow. It’s simply much easier to say ‘we support these percona crds, enjoy’ and it ‘just works’.

**Comment by effespwnz** [comment]
Ingress resource to avoid port forwarding or creating ingress manually/separately?

[a comment by an automated account is not shown]

**Comment by recharte** [comment]
We just released [v1.4.0](https://github.com/percona/everest/releases/tag/v1.4.0) with an official [helm chart](https://github.com/percona/percona-helm-charts/tree/main/charts/everest)
There's also a document with instructions on [how to integrate with ArgoCD](https://github.com/percona/percona-helm-charts/blob/main/charts/everest/docs/argocd.md).

Some of the requested features in this discussion are absent from this initial version but we plan to add them in the future.

**Reply by yurii-olshanetskyi-fox** [reply]
hi folks. I've tried to setup Everest with ArgoCD. I have set application, values. But faced with the issue:

```
version 1.3.0; upgrade.preflightChecks: true/false

Pod: everest-helm-pre-upgrade-hook-780246-7fl45
State: CrashLoopBackOff
State details: back-off 20s restarting failed container=everest-helm-pre-upgrade-hook pod=everest-helm-pre-upgrade-hook-780246-7fl45_everest-system(22ebdee2-c40c-4e47-8243-32b5ea60912b)

Pod Logs: 
Checking requirements for upgrade to version 1.3.0
./everestctl: line 1: Not: not found
```


```
version 1.4.0; upgrade.preflightChecks: true/false

Pod: everest-helm-pre-upgrade-hook-786036-k8mjx
State: CrashLoopBackOff
State details: back-off 40s restarting failed container=everest-helm-pre-upgrade-hook pod=everest-helm-pre-upgrade-hook-786036-k8mjx_everest-system(207952e6-6483-40ac-a6ec-9558d34b3f59)

Pod Logs: 
Checking requirements for upgrade to version 1.4.0
2025-01-17T12:23:33Z	error	output/command.go:64	deployments.apps "percona-everest" not found
could not retrieve Everest version
× deployments.apps "percona-everest" not found
could not retrieve Everest version
```

For both the current and previous version deployment failed on the same step when it tried to create the pod. 

I used all the default values except 2:
```
dbNamespace:
  # -- If set, deploy the database operators in `everest` namespace.
  # The namespace may be overridden by setting `dbNamespace.namespaceOverride`.
  enabled: true
  # -- If `dbNamespace.enabled` is `true`, deploy the database operators in this namespace.
  namespaceOverride: db_test

upgrade:
  preflightChecks: flase
```

**Reply by mayankshah1607** [reply]
Hi @yurii-olshanetskyi-fox,

Before I look into this problem in detail, I just want to check if you had a chance to see our ArgoCD guide: https://github.com/percona/percona-helm-charts/blob/main/charts/everest/docs/argocd.md

There are some known limitations, and you'll need to configure your application a certain way to make it work with our Helm chart.

**Reply by yurii-olshanetskyi-fox** [reply]
hi @mayankshah1607, yes, I used the doc you've provided as guidance. 

my patch-everest.yaml in arocd looks like:

```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: percona-everest
spec:
  destination:
    namespace: everest-system
  source:
    helm:
      valueFiles:
        - values.yaml
        - env/qa/values.yaml
  ignoreDifferences:
    - group: ""
      jsonPointers:
        - /data
      kind: Secret
      name: everest-jwt
      namespace: everest-system
    - group: ""
      jsonPointers:
        - /data
      kind: Secret
      name: everest-accounts
      namespace: everest-system
    - group: ""
      jsonPointers:
        - /data
      kind: Secret
      name: packageserver-service-cert
      namespace: everest-olm
    - group: apiregistration.k8s.io
      jqPathExpressions:
        - .spec.caBundle
        - .metadata.annotations
      kind: APIService
      name: v1.packages.operators.coreos.com
  syncPolicy:
    syncOptions:
      - CreateNamespace=true
      - RespectIgnoreDifferences=true
      - ServerSideApply=true
```

**Reply by mayankshah1607** [reply]
It looks like you have a typo in your values file:
```
upgrade:
  preflightChecks: flase
```
`flase->false`

Please also note that setting this to `true` will simply not work with ArgoCD. Otherwise it tries to run the upgrade hook on every sync which doesn't translate well in GitOps.

Let me know if you have any more issues after fixing the typo. 🙂

**Reply by yurii-olshanetskyi-fox** [reply]
@mayankshah1607  thank you very much, I'm so embarrassing about it  )))

However, fixing the value didn't change the result:

pod: everest-helm-pre-upgrade-hook-455467-skc5p
State: CrashLoopBackOff
log: 
```
Checking requirements for upgrade to version 1.4.0
2025-01-20T16:00:25Z	error	output/command.go:64	deployments.apps "percona-everest" not found
could not retrieve Everest version
× deployments.apps "percona-everest" not found
could not retrieve Everest version
```

![Uploading Screenshot 2025-01-20 at 11.05.37.png…]()

I deleted and created an application from scratch to avoid collisions. However, looks like sth overrides my changes of upgrade.preflightChecks.

UPD:
I've tried to set values directly as helm.values (used copy-paste from github this time) instead of values file. Doesn't matter. ArgoCD is starting pre-upgrade checks.

<img width="764" alt="image" src="https://github.com/user-attachments/assets/be6e2d72-1a6d-4544-8248-382cb6c029c6" />

```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: percona-everest
spec:
  destination:
    namespace: everest-system
  source:
    helm:
      parameters:
        - name: dbNamespace.enabled
          value: "false"
        - name: upgrade.preflightChecks
          value: "false"
#      valueFiles:
#        - env/qa/values.yaml
  ignoreDifferences:
    - group: ""
      jsonPointers:
        - /data
      kind: Secret
      name: everest-jwt
      namespace: everest-system
    - group: ""
      jsonPointers:
        - /data
      kind: Secret
      name: everest-accounts
      namespace: everest-system
    - group: ""
      jsonPointers:
        - /data
      kind: Secret
      name: packageserver-service-cert
      namespace: everest-olm
    - group: apiregistration.k8s.io
      jqPathExpressions:
        - .spec.caBundle
        - .metadata.annotations
      kind: APIService
      name: v1.packages.operators.coreos.com
  syncPolicy:
    syncOptions:
      - CreateNamespace=true
      - RespectIgnoreDifferences=true
      - ServerSideApply=true

```

**Reply by mayankshah1607** [reply]
This is actually a bit strange, because I just tried it and it seems to work fine. All I did was apply the YAML from [here](https://github.com/percona/percona-helm-charts/blob/main/charts/everest/docs/application.yaml) and after a manual sync, everything was up and running:

<img width="366" alt="Screenshot 2025-01-21 at 10 08 22 AM" src="https://github.com/user-attachments/assets/a806c2bb-3f95-4369-9eb7-53f36ecd9ff5" />

I'm not sure what you're doing differently, but I have a few question:
1. What version of the chart are you using?
2. I would have normally known that by looking at the `source` field in your application, but all the details about the chart are missing (`repoURL`,  `chart`, `targetVersion`, etc. Referring to [these](https://github.com/percona/percona-helm-charts/blob/main/charts/everest/docs/application.yaml#L40-L49) lines). So why are these lines missing in your config, and how does the Application know where to use the chart from?
3. What version of ArgoCD are you using?

**Reply by yurii-olshanetskyi-fox** [reply]
hi @mayankshah1607, thank you for your help. We usually use more complex flow through our internal gitlab. But when I use config from [here](https://github.com/percona/percona-helm-charts/blob/main/charts/everest/docs/application.yaml) as is - it runs and everything looks fine. 
Moreover, I have created an ingress config and I can access Everest without patching it services. So Everest application remains in sync:

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: everest-ingress
  namespace: everest-system
spec:
  ingressClassName: traefik
  rules:
    - host: everest.dev.sandbox.com
      http:
        paths:
          - backend:
              service:
                name: everest
                port:
                  number: 8080
            path: /
            pathType: Prefix
```
